examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,635 @@
+    # Generate the embeddings for the images using the xception model.
+    embeddings = xception(xception_input)
+    # Project the embeddings produced by the model.
+    outputs = project_embeddings(
+        embeddings, num_projection_layers, projection_dims, dropout_rate
+    )
+    # Create the vision encoder model.
+    encoder = keras.Model(inputs, outputs, name="vision_encoder")
+    return encoder
+
+
+"""
+## Implement the text encoder
+
+In this example, we use [BERT](https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-12_H-256_A-4/1)
+from [TensorFlow Hub](https://tfhub.dev) as the text encoder
+"""
+
+bert_module_url = (
+    "https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-4_H-512_A-8/1"
+)
+preprocess_model_url = "https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/2"
+
+
+def create_text_encoder(
+    num_projection_layers, projection_dims, dropout_rate, trainable=False
+):
+    # Load the BERT preprocessing module.
+    preprocess = hub.KerasLayer(preprocess_model_url, name="text_preprocessing")
+    # Load the pre-trained BERT model to be used as the base encoder.
+    bert = hub.KerasLayer(bert_module_url, "bert")
+    # Set the trainability of the base encoder.
+    bert.trainable = trainable
+    # Receive the text as inputs.
+    inputs = layers.Input(shape=(), dtype=tf.string, name="text_input")
+    # Preprocess the text.
+    bert_inputs = preprocess(inputs)
+    # Generate embeddings for the preprocessed text using the BERT model.
+    embeddings = bert(bert_inputs)["pooled_output"]
+    # Project the embeddings produced by the model.
+    outputs = project_embeddings(
+        embeddings, num_projection_layers, projection_dims, dropout_rate
+    )
+    # Create the text encoder model.
+    encoder = keras.Model(inputs, outputs, name="text_encoder")
+    return encoder
+
+
+"""
+## Implement the dual encoder
+
+To calculate the loss, we compute the pairwise dot-product similarity between
+each `caption_i` and `images_j` in the batch as the predictions.
+The target similarity between `caption_i`  and `image_j` is computed as
+the average of the (dot-product similarity between `caption_i` and `caption_j`)
+and (the dot-product similarity between `image_i` and `image_j`).
+Then we use crossentropy to compute the loss between the targets and the predictions.
+"""
+
+
+class DualEncoder(keras.Model):
+    def __init__(self, text_encoder, image_encoder, temperature=1.0, **kwargs):
+        super(DualEncoder, self).__init__(**kwargs)
+        self.text_encoder = text_encoder
+        self.image_encoder = image_encoder
+        self.temperature = temperature
+        self.loss_tracker = keras.metrics.Mean(name="loss")
+
+    @property
+    def metrics(self):
+        return [self.loss_tracker]
+
+    def call(self, features, training=False):
+        # Place each encoder on a separate GPU (if available).
+        with tf.device("/gpu:0"):
```

Comment:
The device scopes are hardcoding the assumption that the runtime has 2 GPUs, which is not a safe assumption. We should make this optional.

Besides I'm not sure putting the device scopes in `call` is the right approach wrt variable placement. Consider instead creating the models (e.g. calling `create_vision_encoder`) under the device scope.

Reply:
I tried the code with a machine with no GPUs, 1 GPU, and 2 GPUs, and in all cases the code worked fine. It seems like tf falls back to the available devices if it cannot find the device in the scope. So I think it is safe to leave it as-is.

Besides, I tried to define the scope in the creation:
```
with tf.device("/gpu:0"):
    vision_encoder = create_vision_encoder(
        num_projection_layers, projection_dims, dropout_rate
    )
```
But it didn't work as expected (with 2 GPUs, only one was utilized). Do you mean to define the scope somewhere else?

In fact, in the [Keras FAQ](https://keras.io/getting_started/faq/) regarding model parallelism, the example shows the scope definition around calling the layer/model (btw, the example should be updated to change `tf.device_scope` to `tf.device`)

Let me know your thoughts.

Reply:
Ok, if it works, then sounds good. Please add a comment explaining that TF will fallback on available devices if there are fewer than 2 GPUs.